realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java

```diff
@@ -521,6 +521,99 @@ private void checkEmpty(String str) {
+        boolean checkColumnType = validColumnTypes != null && validColumnTypes.length > 0;
+        if (containsDot(fieldDescription)) {
+            // Resolve field description down to last field name
+            String[] names = fieldDescription.split("\\.");
+            long[] columnIndices = new long[names.length];
+            for (int i = 0; i < names.length - 1; i++) {
+                long index = table.getColumnIndex(names[i]);
+                if (index < 0) {
+                    throw new IllegalArgumentException("Invalid query: " + names[i] + " does not refer to a class.");
+                }
+                RealmFieldType type = table.getColumnType(index);
+                if (type == RealmFieldType.OBJECT || type == RealmFieldType.LIST) {
+                    table = table.getLinkTarget(index);
+                    columnIndices[i] = index;
+                } else {
+                    throw new IllegalArgumentException("Invalid query: " + names[i] + " does not refer to a class.");
+                }
+            }
+
+            // Check if last field name is a valid field
+            String columnName = names[names.length - 1];
+            long columnIndex = table.getColumnIndex(columnName);
+            columnIndices[names.length - 1] = columnIndex;
+            if (columnIndex < 0) {
+                throw new IllegalArgumentException(columnName + " is not a field name in class " + table.getName());
+            }
+            if (checkColumnType && !isValidType(table.getColumnType(columnIndex), validColumnTypes)) {
+                throw new IllegalArgumentException(String.format("Field '%s': type mismatch.", names[names.length - 1]));
+            }
+            return columnIndices;
+        } else {
+            if (getFieldIndex(fieldDescription) == null) {
+                throw new IllegalArgumentException(String.format("Field '%s' does not exist.", fieldDescription));
+            }
+            RealmFieldType tableColumnType = table.getColumnType(getFieldIndex(fieldDescription));
+            if (checkColumnType && !isValidType(tableColumnType, validColumnTypes)) {
+                throw new IllegalArgumentException(String.format("Field '%s': type mismatch. Was %s, expected %s.",
+                        fieldDescription, tableColumnType, Arrays.toString(validColumnTypes)));
+            }
+            return new long[] {getFieldIndex(fieldDescription)};
+        }
+    }
+
+    private boolean containsDot(String s) {
+        return s.indexOf('.') != -1;
+    }
+
+    private boolean isValidType(RealmFieldType columnType, RealmFieldType[] validColumnTypes) {
+        for (int i = 0; i < validColumnTypes.length; i++) {
+            if (validColumnTypes[i] == columnType) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Returns the underlying table backing this schema.
+     *
+     * @return the underlying {@link Table}.
+     */
+    Table getTable() {
```

Comment:
Or we could make `table` package protected and avoid the getter. It's already final so we don't have to worry about setting the value to something else.